src/bin/muxer/avformat.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2017-2018, The rav1e contributors. All rights reserved
+//
+// This source code is subject to the terms of the BSD 2 Clause License and
+// the Alliance for Open Media Patent License 1.0. If the BSD 2 Clause License
+// was not distributed with this source code in the LICENSE file, you can
+// obtain it at www.aomedia.org/license/software. If the Alliance for Open
+// Media Patent License 1.0 was not distributed with this source code in the
+// PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+
+#[cfg(feature = "ffmpeg-sys")]
+use ffmpeg_sys::*;
+
+#[allow(unused_imports)]
+use super::Muxer;
+#[allow(unused_imports)]
+use std::ffi::CString;
+#[allow(unused_imports)]
+use std::io;
+#[allow(unused_imports)]
+use std::mem;
+#[allow(unused_imports)]
+use std::ptr;
+
+#[cfg(feature = "ffmpeg-sys")]
+pub struct AvformatMuxer {
+  context: *mut AVFormatContext,
+  duration: i64
+}
+
+#[cfg(feature = "ffmpeg-sys")]
+impl AvformatMuxer {
+  pub fn open(path: &str) -> Box<dyn Muxer> {
+    unsafe {
+      let mut context = ptr::null_mut();
+      let p = CString::new(path).expect("convert to cstring failed");
+      match avformat_alloc_output_context2(
+        &mut context,
+        ptr::null_mut(),
+        ptr::null(),
+        p.as_ptr() as *const i8
+      ) {
+        0 => {
+          let mut io = ptr::null_mut();
+          let ret =
+            avio_open(&mut io, p.as_ptr() as *const i8, AVIO_FLAG_WRITE);
```

Comment:
Why are we using AVIO directly?

Reply:
for mux, caller need set/close the avio. 

see
https://github.com/FFmpeg/FFmpeg/blob/819ed1df94dff335c2653a83e1d3001aa5797f9e/libavformat/avformat.h#L1374-L1386